pkg/datapath/micro_service_watcher.go

```diff
@@ -0,0 +1,436 @@
+}
+
+func newMicroServiceBRWatcher(client client.Client, kubeClient kubernetes.Interface, mgr manager.Manager, taskType string, taskName string, namespace string,
+	podName string, containerName string, associatedObject string, callbacks Callbacks, log logrus.FieldLogger) AsyncBR {
+	ms := &microServiceBRWatcher{
+		mgr:              mgr,
+		client:           client,
+		kubeClient:       kubeClient,
+		namespace:        namespace,
+		callbacks:        callbacks,
+		taskType:         taskType,
+		taskName:         taskName,
+		thisPod:          podName,
+		thisContainer:    containerName,
+		associatedObject: associatedObject,
+		eventCh:          make(chan *v1.Event, 10),
+		podCh:            make(chan *v1.Pod, 2),
+		wgWatcher:        sync.WaitGroup{},
+		log:              log,
+	}
+
+	return ms
+}
+
+func (ms *microServiceBRWatcher) Init(ctx context.Context, param interface{}) error {
+	ms.ctx, ms.cancel = context.WithCancel(ctx)
+	succeeded := false
+
+	eventInformer, err := ms.mgr.GetCache().GetInformer(ms.ctx, &v1.Event{})
+	if err != nil {
+		return errors.Wrap(err, "error getting event informer")
+	}
+
+	podInformer, err := ms.mgr.GetCache().GetInformer(ms.ctx, &v1.Pod{})
+	if err != nil {
+		return errors.Wrap(err, "error getting pod informer")
+	}
+
+	eventHandler, err := eventInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				evt := obj.(*v1.Event)
+				if evt.InvolvedObject.Namespace != ms.namespace || evt.InvolvedObject.Name != ms.associatedObject {
+					return
+				}
+
+				ms.log.Infof("Pushed adding event %s/%s, message %s for object %v", evt.Namespace, evt.Name, evt.Message, evt.InvolvedObject)
+
+				ms.eventCh <- evt
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				evt := obj.(*v1.Event)
+				if evt.InvolvedObject.Namespace != ms.namespace || evt.InvolvedObject.Name != ms.associatedObject {
+					return
+				}
+
+				ms.log.Infof("Pushed updating event %s/%s, message %s for object %v", evt.Namespace, evt.Name, evt.Message, evt.InvolvedObject)
+
+				ms.eventCh <- evt
+			},
+		},
+	)
+
+	if err != nil {
+		return errors.Wrap(err, "error registering event handler")
+	}
+
+	defer func() {
+		if !succeeded {
+			if err := eventInformer.RemoveEventHandler(eventHandler); err != nil {
+				ms.log.WithError(err).Warn("Failed to remove event handler")
+			}
+		}
+	}()
+
+	podHandler, err := podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(_, obj interface{}) {
+				pod := obj.(*v1.Pod)
+				if pod.Namespace != ms.namespace || pod.Name != ms.thisPod {
+					return
+				}
+
+				if pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed {
+					ms.podCh <- pod
+				}
+			},
+		},
+	)
+
+	if err != nil {
+		return errors.Wrap(err, "error registering pod handler")
+	}
+
+	defer func() {
+		if !succeeded {
+			if err := podInformer.RemoveEventHandler(podHandler); err != nil {
```

Comment:
Shouldn't we use `ms.Close()` here ? Are we not missing the `cancel()` call as well ? 
same as line 148 comment earlier. Curious to know if my understanding is incorrect.

Reply:
or do we need to combine both the defer calls for podhandler and eventhandler and then call `ms.Close()`

Reply:
See https://github.com/vmware-tanzu/velero/pull/7999#discussion_r1690716165